# Review   - 2,000xp



## Preliminaries: 1000xp
Here we create a list of our friends:


In [1]:
friends = ['Ann', 'Ben', 'Clara', 'David', 'Kelly', 'Rachel']

Can you write a loop that will print out each of our friends? Something like this:

    Ann
    Ben
    Clara
    David
    Kelly
    Rachel

In [46]:
# Your code here

Suppose we have a list of numbers:


In [4]:
numbers = [11, 22, 37, 51, 17]

We want to print the sum of the numbers but we have an error in the following code. Can you fix it?

In [5]:
total = 0
for num in numbers:
    total = total + num
print(num)

17


We have a list of grades:

In [6]:
grades = [89, 91, 79, 85, 97, 91]


Can you print out each grade, one per line like this:

    89
    91
    79
    85
    97
    91
    

Now we want to print out only the grades over 90:

    91
    97
    91

Can you change your code to do so?

We represent a sentence as a list of words.

In [7]:

sentence = ['the', 'man', 'saw', 'the', 'girl', 'on', 'the', 'hill', 'with', 'the', 'telescope']

Can you write code that will compute and print the number of occurrences of the word *the* in the sentence?

In [47]:
# Your code here

We have another sentence

In [11]:
sentence2 = ['the', 'man', 'saw', 'a', 'girl', 'on', 'the', 'hill', 'with', 'the', 'telescope', 'on', 'a', 'sunny', 'day']

and we want to count the occurrences of *the* and *a* and print something like:

    the: 4
    a:   2
    
Can you write the code?

We have a list of our friends and their home states:

In [12]:
friends = [['Ann', 'North Dakota'], ['Ben', 'Virginia'], ['Clara', 'Connecticut'], 
           ['David', 'Minnesota'], ['Kelly', 'Kansas'], ['Rachel', 'New Mexico']]

Plus we set a variable for the person we want to search for:

In [13]:
searchPerson = 'Kelly'

Now we want to write code that finds the home state for the person `searchPerson` and prints it. For example, if the value of `searchPerson` is `Kelly`  the code should print out `Kansas`. If `searchPerson` is `Rachel` it should print out `New Mexico`. Test your code by changing the value of searchPerson above.

In [48]:
# Your code here

Now we have a list of friends and their ages:

In [17]:
friends = [['Ann', 21], ['Ben', 17], ['Clara', 24], 
           ['David', 27], ['Kelly', 18], ['Rachel', 20],
           ['Jennifer', 15], ['Sam', 19], ['Vivic', 16]]

We would like to print out the friends who are between the ages of 18 and 21 inclusive (meaning if someone is 18 that person's name should be printed)

In [49]:
# Your work here

# Dataframes
<img src="http://www2.pictures.zimbio.com/gi/Artistic+Gymnastics+World+Championships+2009+J2OThe0-1Wul.jpg" />


Let's load in a small dataset of women athletes:

In [19]:
import pandas as pd
athletes = pd.read_csv('https://raw.githubusercontent.com/zacharski/machine-learning/master/data/athletes.csv')
athletes.head()

Name       Sport  Height  Weight
0    Asuka Teramoto  Gymnastics      54      66
1  Brittainey Raven  Basketball      72     162
2          Chen Nan  Basketball      78     204
3     Gabby Douglas  Gymnastics      49      90
4  Helalia Johannes       Track      65      99

The athletes either compete professionally in gymnastics, basketball (WNBA) or track (running marathons).

Remember when you worked on the Titanic problem and had to come up with a rule to decide whether a person survived or not?  Your code looked something like this:


    def predict(passenger):
    # Your code here
    # You should return 1 if you think the person survived and 0 if you think they didn't
    # For example, if you think males survived but females didn't you would write:
    if passenger['Sex'] == 'female' and passenger['Pclass'] != 3 :
        return 1
    elif passenger['Age'] < 5:
        return 1
    else:
        return 0

I would like you to write similar code, but this time to predict whether someone is a gymnast, a basketball player, or runs track. (Your code should return one of the strings `Gymnastics, Basketball, ` or `Track`. 
You should only use the attributes Height and Weight. Here is the code to alter:

In [50]:
def predict(athlete):
    # Your code here.
    # to get the height of the person you would use athlete['Height']
    # For now we will always predict 'Basketball' (You need to change this)
    
    return 'Basketball'

Let's run our code and see what our accuracy is:




In [51]:
testdatafile = 'https://raw.githubusercontent.com/zacharski/machine-learning/master/data/athletes-gbv.csv'
testdata = pd.read_csv(testdatafile)
predictions = []
totalCorrect = 0
for athlete_index, athlete in testdata.iterrows():
    predictions.append(predict(athlete))

    
from sklearn.metrics import accuracy_score
score = accuracy_score(testdata['Sport'], predictions)    

    
print('Accuracy %5.3f' % (score * 100))




Accuracy 33.333


Since the default code always predicts `Basketball` and for our test set we have, for each sport, 25 players, our accuracy is 33.333%.  Can you do better?

### XP

% accuracy | xp
:---: | :---:
< 50 | 250
50 - 79 | 500 + (accuracy - 50) x 16.5
>=80    | 1000 + (accuracy - 80) x 50